### PR TITLE
Introduce JUnit5 as testing backend for CLI tests

### PR DESCRIPTION
**Pull request by StefanSpieker**
I just tried out how hard it would be to provide junit5 support in parallel to junit4 and I changed the CLI tests to use junit5 and used two features which might be useful also for other test classes:

> assertThrows()

and 

> @Execution(ExecutionMode.CONCURRENT)

A guide for migration can be found here: https://junit.org/junit5/docs/current/user-guide/#migrating-from-junit4

Do you think junit5 is useful at all for jenkins? 

### Proposed changelog entries

* Internal: Use junit5 for CLI tests

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@oleg-nenashev: What do you think about that?

### CODE REVIEW

File: cli/src/test/java/hudson/cli/PrivateKeyProviderTest.java [hunk 1]
```diff
@@ -84,21 +89,21 @@ public void loadKeyOpenSSH() throws IOException, GeneralSecurityException {
     /**
     key command: ssh-keygen -f openssh-unsupported -t rsa -b 1024 -p password
     */
-    @Test(expected = NoSuchAlgorithmException.class)
+    @Test
```
Comment (MRamonLeon) [inline cli/src/test/java/hudson/cli/PrivateKeyProviderTest.java]: is this change required by the new jUnit version? Because it may break a lot of tests in this case.

**Review by rsandell — APPROVED** [review]
Looks good to me, but do we really need it?

File: cli/src/test/java/hudson/cli/PrivateKeyProviderTest.java (hunk 1 above)
Comment (rsandell) [reply cli/src/test/java/hudson/cli/PrivateKeyProviderTest.java]: it's replaced by the `assertThrows`

**Comment by MRamonLeon** [comment]
Changes and [documentation](https://junit.org/junit5/docs/current/user-guide/#migrating-from-junit4) on jUnit website looks good to have both versions living together, but better to have more eyes on it.

**Comment by rsandell** [comment]
Maybe start a thread on the dev list asking for opinions to include this throughout core?

Though I'm a bit hesitant to change tests just for changing it. It can throw off reports and what not.

But introducing it for new tests I'm way OK with :)

**Review by timja — COMMENTED** [review]
What's the reason to do it in CLI and not all of core?

**Comment by StefanSpieker** [comment]
I would definitly integrate this in core. But I thought CLI would be an easy start and since it just has a few tests porting them to junit 5 as a demonstration was quite easy.

**Review by MarkEWaite — APPROVED** [review]
+1 from me.  Test automation improvement is a good thing.

I think we'll want in either the [plugin BOM](https://github.com/jenkinsci/bom) or the Jenkins BOM so that plugins can have the version information provided to them.

**Review by res0nance — APPROVED** [review]
Seems ok for CLI but from the docs I read it seems to not have full support for Rules this might take some pretty heavy migration work for plugins to have jUnit5

**Comment by MarkEWaite** [comment]
> Seems ok for CLI but from the docs I read it seems to not have full support for Rules this might take some pretty heavy migration work for plugins to have jUnit5

I migrated 4 of 8 tests of the platformlabeler from JUnit 4 to JUnit 5 with https://github.com/jenkinsci/platformlabeler-plugin/pull/115.  The other 4 tests use JenkinsRule.  Parameterized tests are considered "experimental", and that was 3 of the 4 converted tests.  The tests seem well behaved with no detected differences in test reporting or test coverage from my Jenkins jobs.  The platformlabeler JUnit 5 tests run with Java 8 and with Java 11 and with older Jenkins versions (2.138.4) and newer Jenkins versions (2.164.3 and 2.176.3).

I don't object to this being merged, but I agree that without something like JenkinsRule, I would expect to see very limited adoption.

**Comment by rsandell** [comment]
I am a bit hesitant that we change all old tests over to JUnit 5 as there is a risk reproduced bugs might not be reproduced any more. Though changing a few just to prove that JUnit 5 works is OK, and adding it for future tests to use is great.

**Review by fcojfernandez — APPROVED** [review]

**Review by timja — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
Looks like there is a consensus here.
Unless somebody is against the change, will go ahead and merge it this week
